Add quadraturefunctions to visit datacollection [quadspace-visit-dev]

IdoAkkerman opened the pull request:
This PR adds  quadraturefunctions to the visit data collection.
At this stage the quadraturefunctions can be written to file and read again.
Visualisation capability in visit will be worked on in parallel.

See #1477.

<!--GHEX{"id":1513,"author":"IdoAkkerman","editor":"tzanio","reviewers":["rcarson3","bslazarov"],"assignment":"2020-05-31T16:30:22-07:00","approval":"2020-07-03T04:38:44.636Z","merge":"2020-07-10T04:38:44.636Z"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1513](https://github.com/mfem/mfem/pull/1513) | @IdoAkkerman | @tzanio | @rcarson3 + @bslazarov | 05/31/20 | 07/02/20 | ⌛due 07/09/20 | |
<!--ELBATXEHG-->

tzanio commented:
LGTM

kennyweiss commented on fem/datacollection.cpp:
```diff
@@ -598,14 +605,28 @@ void VisItDataCollection::LoadFields()
       if (serial)
       {
-         field_map.Register(it->first, new GridFunction(mesh, file), own_data);
+         if ((it->second).association == "nodes")
```
What are the valid values for ``VisItFieldInfo.association``?
Looking at ``RegisterField`` and ``RegisterQField``, this looks right.

Perhaps some comments can be added here?
https://github.com/mfem/mfem/blob/d87f4e347c1332181263a8f90adb6e66db479d01/fem/datacollection.hpp#L388-L397

IdoAkkerman replied:
In Visit there is already a hook for the `elements` instead of `qpoints` I introduced here.
So I think it might be better to adopt `elements`, which is currently not used, as far as I know.

IdoAkkerman replied:
There is also the issue of the LOD. I would like to use that to tell Visit the integration order.
Or should I introduce a new parameter in the file to  pass this info to visit.
Then there is also the issue of mapping an integration routine with the appropriate refinement routine in visit. I am aiming for a minimal implementation with abundant checks for cases that will not work.

IdoAkkerman commented:
I have added `GetRefinementLevelFromElems` to translate the number quad points to an LOD.
This gives a -1 if not match is found. In that case visit will display an element average instead of the individual quadrature points.

I added a `openGL` point distribution options, this allows the GL points to be beter visualised in Visit. Instead of a uniform subdivision of elements, the subdivision is governed by the distribution of quad points.

rcarson3 commented:
@IdoAkkerman could you provide us with a relatively simple example to test this on at least for the PR? Maybe something like where the quadrature data you're saving off is a projection of the mesh nodes on the quadrature points (a single direction of the project is fine as well so we can plot a scalar field in visit). You could probably use these lines as a starting point https://github.com/mfem/mfem/blob/master/tests/unit/fem/test_quadf_coef.cpp#L20-L72

v-dobrev commented on fem/datacollection.cpp:
```diff
@@ -431,6 +428,27 @@ void VisItDataCollection::RegisterField(const std::string& name,
+   Mesh *mesh = qf->GetSpace()->GetMesh();
+   for (int e=0; e<qf->GetSpace()->GetNE(); e++)
+   {
+      int locLOD = GlobGeometryRefiner.GetRefinementLevelFromElems(
+                      mesh->GetElementBaseGeometry(e),
+                      qf->GetElementIntRule(e).GetNPoints());
```
Why not just use `IntegrationRule::GetOrder` to derive the LOD?

IdoAkkerman replied:
There is unfortunately not a (very) clear connection between quadrature order and quadrature points. Moreover not each distribution of quadrature points is associated with a refinement, at least not how the refinements are programmed at this time.  Off course some of this "logic" could also be moved to visit, and LOD could simply be the order (or order +1) similar as for the FESpace,

IdoAkkerman replied:
Moreover the quadrature order is ussually ~ twice as high as the order of the shapefunctions.

IdoAkkerman commented:
@rcarson3 i have done that for debug purposes. But will add a case to the unit tests, this should also simplify debuging/testsuites on the VisIt side. I will work probably not have time to work on it  before the 22nd.

IdoAkkerman commented:
@rcarson3 Was able to pulls something together in the last minute.

IdoAkkerman commented:
Will de AFK till the 23th.

rcarson3 commented on tests/unit/fem/test_quadf_coef.cpp:
```diff
@@ -156,6 +157,73 @@ TEST_CASE("Quadrature Function Coefficients",
       REQUIRE(output.Norml2() < tol);
    }
 
+   SECTION("Write and Read Scalar Quadraturefunction")
```
So the changes made to this unit test really don't belong in this file. They would be better suited in the /tests/unit/fem/test_datacollection.cpp file where the other data collection are. So, I would suggest reverting changes made here and moving them over to the other file.

IdoAkkerman replied:
Indeed don't know why I did not consider that. Will move, and change the test to better fit the datacollection unittest.

rcarson3 commented on tests/unit/CMakeLists.txt:
```diff
@@ -18,35 +18,6 @@ include_directories(BEFORE ${PROJECT_BINARY_DIR})
 include_directories(BEFORE ${CMAKE_CURRENT_SOURCE_DIR})
 
 set(UNIT_TESTS_SRCS
```
Why are all of these deleted? Please revert these changes back.

IdoAkkerman replied:
Sorry that was sloppy of me... just deleted them to speed up testing.

rcarson3 commented on fem/intrules.hpp:
```diff
@@ -293,7 +294,8 @@ class Quadrature1D
       ClosedUniform   = 3,  ///< aka closed Newton-Cotes
-      OpenHalfUniform = 4   ///< aka "open half" Newton-Cotes
+      OpenHalfUniform = 4,  ///< aka "open half" Newton-Cotes
+      ClosedGL        = 5   // Closed Gauss Legendre
```
Convert to a doxygen comment and make comment inline with the style used in the above comments

IdoAkkerman replied:
Will do so.

rcarson3 commented on fem/fespace.hpp:
```diff
@@ -723,6 +723,12 @@ class QuadratureSpace
    /// Return the total number of quadrature points.
    int GetSize() const { return size; }
 
+   /// Returns the mesh
+   inline Mesh *GetMesh() const { return mesh; }
+
+   /// Returns number of elements in the mesh.
+   inline int GetNE() const { return mesh->GetNE(); }
+
```
These should be nice improvements to life functions to ```QuadratureSpace```.

IdoAkkerman replied:
I do not understand the comment? I interpret this as positive feadback and not a request for change. Is that correct?
If so, thank you! If not could you eleborate?

rcarson3 replied:
It was just positive feedback :) and nothing for you to fix.

rcarson3 commented on fem/datacollection.hpp:
```diff
@@ -445,6 +446,10 @@ class VisItDataCollection : public DataCollection
    virtual void RegisterField(const std::string& field_name, GridFunction *gf);
 
+   /// Add a quadrature function to the collection and update the root file
+   virtual void RegisterQField(const std::string& q_field_name,
```
Another comment should be added that this currently is not usable within Visit to visualize quadrature data.

Based on the below unit test, if you do try to within Visit it fails with an error from MFEM saying: ```MFEM abort: input stream is not a FiniteElementSpace!```

IdoAkkerman replied:
Will make a statement that VisIt 3.12 does not work. A patch has been sent to VisIt developers.

IdoAkkerman commented:
@rcarson3 thank you for your input for correctting the sloppyness and making improvements.  I made the changes as requested.

rcarson3 commented on fem/datacollection.hpp:
```diff
@@ -445,6 +446,12 @@ class VisItDataCollection : public DataCollection
    virtual void RegisterField(const std::string& field_name, GridFunction *gf);
 
+   /// Add a quadrature function to the collection and update the root file.
+   /** Visualisation of quadrature function is not supported in VisIt(3.12).
```
typo here should be Visualization

tzanio commented:
Extending the deadline to this Friday...

tzanio commented:
Merged in `next` for testing...

tzanio commented:
Merged in `next` for testing...
